### PR TITLE
Benchmark Wrappers

### PR DESCRIPTION
**Pull request by WilsonGregory**
## Changes
- a mixed bag of different changes related to the benchmark capabilities, more still to come
- made BenchmarkableModel, a class that provides the getBenchmarker function, and expects there to be a benchmarkerFunctional function in the inheriting subclass
- added LassoNetWrapper, a class that wraps LassoNetClassifier and implements benchmarkerFunctional
- added RandomBaseline, a random baseline model that just picks k random indices for the markers
- implemented benchmarkerFunctional for Global Gate and Concrete VAE.
- add a mode to plot_benchmarks to allow specifying misclass rate, accuracy, and later other types of output

## Tests
- ran the benchmark script
- add to tests for plot_benchmarks

## Doc Changes
- none

## Future TODOs:
- finish implementing benchmarkFunctional for smashpy, L1 VAE, RankCorr
- return test_rep, cm from the benchmark function in a way they can be reasonably used

### CODE REVIEW

File: tests/test_plots.py [hunk 1]
```diff
@@ -49,3 +49,15 @@ def test_two_models_5_tries(self):
...
     with plt.ion():
       plot_benchmarks(misclass_rates, 'k', benchmark_range=[10, 25, 50, 100])
+
+  def test_modes(self):
+    misclass_rates = {
+      'Supervised Marker Map': np.array([[0.18801997, 0.23460899]]),
+      'Mixed Marker Map': np.array([[0.19966722, 0.26622296]])
+    }
+
+    with plt.ion():
```
Comment (beelze-b) [inline tests/test_plots.py:59]: What does plt.ion do here? Is this `with` so that the plots are updated as these, possibly long running, benchmarks complete?

File: src/utils.py [hunk 2]
```diff
@@ -1261,29 +1416,40 @@ def benchmark(models, num_times, X, y, benchmark, train_size = 0.7, val_size = 0
         )
 
         #recombine train and val for methods that don't use a val set, and for training model after finding markers
```
Comment (beelze-b) [inline src/utils.py:1418]: I think these comments need to be removed in the new stuff.

I am a bit confused about the train and validation stuff since some of the methods that we compare to don't make a distinction between train and validation sets when forming the models that are proxies to evaluate the markers. 

Can you explain a bit this change?

File: src/utils.py [hunk 3]
```diff
@@ -1261,29 +1416,40 @@ def benchmark(models, num_times, X, y, benchmark, train_size = 0.7, val_size = 0
...
-        X_train = X[np.concatenate([train_indices, val_indices]), :]
-        y_train = y[np.concatenate([train_indices, val_indices])]
+        X_train = X[train_indices, :]
+        y_train = y[train_indices]
+        X_val = X[val_indices, :]
+        y_val = y[val_indices]
+        X_train_val = X[np.concatenate([train_indices, val_indices]), :]
+        y_train_val = y[np.concatenate([train_indices, val_indices])]
         X_test = X[test_indices,:]
         y_test = y[test_indices]
```
Comment (beelze-b) [inline src/utils.py:1427]: I think these comments need to be removed in the new stuff.

I am a bit confused about the train and validation stuff since some of the methods that we compare to don't make a distinction between train and validation sets when forming the models that are proxies to evaluate the markers.

Can you explain a bit this change?

File: tests/test_plots.py [hunk 4]
```diff
@@ -49,3 +49,15 @@ def test_two_models_5_tries(self):
...
+      plot_benchmarks(misclass_rates, 'k', mode='misclass', benchmark_range=[10,25])
+      plot_benchmarks(misclass_rates, 'k', mode='accuracy', benchmark_range=[10,25])
+      with pytest.raises(Exception):
+        plot_benchmarks(misclass_rates, 'k', mode='fake_mode', benchmark_range=[10,25])
```
Comment (beelze-b) [inline tests/test_plots.py:63]: What does plt.ion do here? Is this with so that the plots are updated as these, possibly long running, benchmarks complete?

File: src/utils.py (hunk 3 above)
Comment (beelze-b) [reply src/utils.py:1427]: The comment that goes 

#recombine train and val for methods that don't use a val set, and for training model after finding markers

**Comment by beelze-b** [comment]
I think these fixes are a good way to abstract away the benchmarks. I approve. I have a few questions related to new functionality above, but nothing blocking.

**Review by beelze-b — APPROVED** [review]

File: tests/test_plots.py (hunk 4 above)
Comment (WilsonGregory) [reply tests/test_plots.py:63]: The plot functions call `plt.show` which blocks further python execution until you close the little matplotlib windows that are generated. Since we are running a large battery of tests here, this means we would have to manually close the window for each call to plot_XXX. For these tests, I am just checking that the function runs without breaking (or breaks when it should, i.e. for mode="fake_mode"), rather than actually checking that the output plot looks correct.

I found the `plt.ion()` workaround, it makes it so that the `plt.show` is not blocking. So when I run the tests, python starts to generate the plots, but then the script ends and they are all closed.

Comment (beelze-b) [reply tests/test_plots.py:63]: That makes a lot of sense. Thanks!

File: src/utils.py (hunk 3 above)
Comment (WilsonGregory) [reply src/utils.py:1427]: Ah good catch, yeah that comment isn't accurate anymore. 

For the data splitting, I am glad you brought it up because it is a little wacky how I am doing it right now, so let me know if you think there are better ways of handling it. Basically the complication stems from the fact that I want to call the data splitting in the outermost loop so that all the models use the same data splits. But then the different models use different parts of the data.

So LassoNet uses X_train, y_train, X_val, and y_val. Then the VAE family of models use train_dataloader and val_dataloader. I haven't done the Smash wrapper yet, but that one is even worse because it needs the annotations. So here I was thinking, just pass all the stuff to the benchmarkerFunctional, and then in that function for each model, just use the ones you need.

The X_train_val and y_train_val is just passed to the new_model_metrics. I think I might move that into the function call to make it less confusing.

**Comment by WilsonGregory** [comment]
> I was curious if we have ran some of these benchmarks already because I wanted to give a small warning related to pytorch(lightning) versions. I remember that the pytorch versions need to be somewhat more recent to get the best results from MarkerMap. I can't definitely remember why, but I think it related to some default initialization. Our methods performed at or slightly better to LassoNet in older versions and had a decent advantage in the more recent versions. The results I had uploaded on Github before were ran with the more recent versions. I am not sure if I updated the arxiv paper with the latest results.

Ooh, thanks for the heads up. I have been using the latest pytorch-lightning, 1.5.10. I have had some trouble getting results that look similar to the paper. Here are some benchmarks that I ran:
image
image

Both of these were with the Zeisel data set, and not a huge number of trials (<7). I need to run on Paul and see if it is similar.

Also, this one had Smash RF doing much better than in the paper, but again it is only on 3 runs.
image

This is the plot that was in the paper that I was comparing against: 
image

I haven't tuned any of the hyper-parameters, I have just been using what was in the Zeisel notebook I believe. So that could also be a factor that I am not using the correct hyper-parameters.

File: src/utils.py (hunk 3 above)
Comment (beelze-b) [reply src/utils.py:1427]: Yeah RankCorr and SmashPy have the same issue. 


> and then in that function for each model, just use the ones you need.

Yeah by using some classmethod or a partial or wtv. I think that could work.

Sounds good to me!

**Comment by beelze-b** [comment]
I haven't changed the hyperparameters much.

Another thing might be the pytorch lightning because of how it sets the automatic learning rate. If we really need to, we can load and evaluate the saved model weights and train/test splits on Colab to confirm the previous results. It would be imperative to be able to replicate results on all machines though. 

Zeisel was always the dataset that LassoNet actually beat us often, so I am not surprised at those results. If MarkerMap doesn't outperform MouseBrain and Cite-Seq datasets, then I will be a lot more alarmed.

**Comment by beelze-b** [comment]
I might try to run Supervised MarkerMap zeisel over night on my computer and see what results I get

for k = 50 markers. 

I used to get lower values for classification before so let me see if the irreplicability is a machine / version issue.

**Comment by WilsonGregory** [comment]
> Zeisel was always the dataset that LassoNet actually beat us often, so I am not surprised at those results. If MarkerMap doesn't outperform MouseBrain and Cite-Seq datasets, then I will be a lot more alarmed.

Oh okay, that is good to hear. I will run it on some of the other data sets.

**Comment by beelze-b** [comment]
I think I have good news.

I just made a new branch off the main right before you did the merge.

I ran just markermap supervised on zeisel.

I got the lower misclassification rates that I have been getting before.

Here is a screenshot.

image


Everything else is commented out because I didn't run those models locally. I think the issue might be versions.


I ran this off the following torch and lightning versions.

image

I am quite pleased that I was able to replicate the very good Zeisel results for MarkerMap supervised, which means maybe we can get them replicated by you too. Not gonna lie I was kinda panicking that my previous results were a fluke!

I actually never ran anything on this machine before so that makes me even less panicked because it is fresh!

**Comment by beelze-b** [comment]
Also since you are reading this, not sure if you saw this paper but it helped me out when i was learning about the gumbel max trick!

https://arxiv.org/abs/2110.01515

**Comment by WilsonGregory** [comment]
@beelze-b I am seeing some similar results that you are for this, this was over 10 runs on zeisel:
image
And here is the versions of pytorch:
image
I think these results are similar enough that it is within the randomness of the splits. I am going to rerun the benchmark models to see how it compares to LassoNet and Smash, but I think for k=50 I will be able to reproduce your results.
